### ISSUE TITLE
Unsubscribe by RegExp

### ISSUE THREAD

**Issue by vdakalov** [hemerajs/hemera#229]
https://github.com/hemerajs/hemera/blob/d13b52980cbf947aaa88696f9dc9afeb4e2db6b3/packages/hemera/lib/index.js#L1027
There wrote "if topic is string then take sid by its topic" but "else unsubscribe by topic which does not is string". And if `topic` is not a string, then it is a RegExp object. Result:

```
[11605] 2018/11/20 14:02:39.579853 [TRC] 127.0.0.1:50742 - cid:1 - ->> [UNSUB /^[a-zA-Z0-9-.]+$/i]
```

https://github.com/hemerajs/hemera/blob/d13b52980cbf947aaa88696f9dc9afeb4e2db6b3/packages/hemera/lib/index.js#L1018
This line should be process always

This unsubscribe method does not work:

![nats unsubscribe gif](http://ecoplastpet.ru/nats.gif)

**Comment by StarpTech** [comment]
Hi @vdakalov 

> There wrote "if topic is string then take sid by its topic" but "else unsubscribe by topic which does not is string". And if topic is not a string, then it is a RegExp object. Result:

Where did you read it?

RegExp aren't supported only the topic or the subscription-id.

**Comment by vdakalov** [comment]
Maybe You meant it:
```javascript
    if (typeof topic === 'string' || topic instanceof RegExp) {
```
?

**Comment by StarpTech** [comment]
@vdakalov now I got it. Your topic is from type RegExp and you want to remove it. This wasn't considered yet good catch! Would you mind to create a PR? I have some hints.

1. When the topic is added we have to use the string form like `this._topics.set((new RegExp('a+b+c').toString(), sid)`
2. When the topic is removed you have to use `this._router.remove({ topic: RegExp() })` https://github.com/mcollina/bloomrun/blob/master/test.js#L169

**Comment by StarpTech** [comment]
@vdakalov forget what I said. We can't use a RegexExp as a NATS subject. Here you can find all possible variants https://nats.io/documentation/internals/nats-protocol/.

Would you mind to create a PR which throws an error when no string or number was passed?

**Comment by vdakalov** [comment]
Let me describe steps for reproduce the problem here (PR will be unnecessary)

First, you need start nats server with _trace_ option
```
gnatsd --trace
```
Next just execute follow script
```typescript
import * as nats from 'nats';
import * as Hemera from 'nats-hemera';

const hemera = new Hemera(nats.connect(), {});
hemera.ready(() => {
 // add new subscription
  hemera.add({
    topic: '>' // <--- full wildcard topic
  }, (request, callback) => callback(null, 'Cut the Rope'));
  // send request
  hemera.act({
    topic: 'UndefinedTopic' // <--- this topic does not exist
  }, (error, response) => {
    console.log('Who are you: ' + response)
    hemera.close(); // <-- REQUIRED!
  });
});
```
Now look trace logs
```
...
[923] 2018/11/21 08:19:42.207369 [TRC] 127.0.0.1:55560 - cid:1 - ->> [UNSUB /^[a-zA-Z0-9-.]+$/i]
...
```

## What happen?

1. Since topic for add is a full-wildcard token, then it tranclated to RegExp, here:
https://github.com/hemerajs/hemera/blob/d13b52980cbf947aaa88696f9dc9afeb4e2db6b3/packages/hemera/lib/index.js#L1104
And this RegExp object set to `this._topics` Map registry as key, here
https://github.com/hemerajs/hemera/blob/d13b52980cbf947aaa88696f9dc9afeb4e2db6b3/packages/hemera/lib/index.js#L1146
2. When calling of `hemera.close` it will make `removeAll` subscriptions, here:
https://github.com/hemerajs/hemera/blob/d13b52980cbf947aaa88696f9dc9afeb4e2db6b3/packages/hemera/lib/index.js#L1532
3. Finally, it call `remove` function with RegExp object as topic argument.
https://github.com/hemerajs/hemera/blob/d13b52980cbf947aaa88696f9dc9afeb4e2db6b3/packages/hemera/lib/index.js#L1533

**Comment by StarpTech** [comment]
Hi @vdakalov thanks for the nice explanation. RegExp must be handled differently in comparison and if we want to remove it from the index. The PR is still needed. Would you like to create one?

**Comment by vdakalov** [comment]
You're welcome. Sorry, I have no time :( You can?

**Comment by vdakalov** [comment]
https://github.com/mcollina/bloomrun/issues/48#issuecomment-441600349

**Comment by StarpTech** [comment]
Closed in #236
